mycroft/skills/skill_manager.py

```diff
@@ -464,57 +507,102 @@ def _remove_git_locks(self):
-        LOG.info('Loading internet skills...')
+        LOG.info('Loading skills that require internet...')
         self._load_new_skills(network=True, internet=True)
         self._internet_loaded.set()
 
+    def _unload_on_network_disconnect(self):
+        """ unload skills that require network to work """
+        for skill_dir in self._get_skill_directories():
+            # by definition skill_id == folder name
+            skill_id = os.path.basename(skill_dir)
+            skill_loader = self._get_skill_loader(skill_dir, init_bus=False)
+            requirements = skill_loader.runtime_requirements
+            if requirements.requires_network and \
+                    not requirements.no_network_fallback:
+                # unload until network is back
+                self._unload_skill(skill_dir)
+
+    def _unload_on_internet_disconnect(self):
+        """ unload skills that require internet to work """
+        for skill_dir in self._get_skill_directories():
+            # by definition skill_id == folder name
+            skill_id = os.path.basename(skill_dir)
+            skill_loader = self._get_skill_loader(skill_dir, init_bus=False)
+            requirements = skill_loader.runtime_requirements
+            if requirements.requires_internet and \
+                    not requirements.no_internet_fallback:
+                # unload until internet is back
+                self._unload_skill(skill_dir)
+
+    def _unload_on_gui_disconnect(self):
+        """ unload skills that require gui to work """
+        for skill_dir in self._get_skill_directories():
+            # by definition skill_id == folder name
+            skill_id = os.path.basename(skill_dir)
+            skill_loader = self._get_skill_loader(skill_dir, init_bus=False)
+            requirements = skill_loader.runtime_requirements
+            if requirements.requires_gui and \
+                    not requirements.no_gui_fallback:
+                # unload until gui is back
+                self._unload_skill(skill_dir)
+
     def _load_on_startup(self):
         """Handle initial skill load."""
         LOG.info('Loading offline skills...')
         self._load_new_skills(network=False, internet=False)
 
-    def _load_new_skills(self, network=None, internet=None):
+    def _load_new_skills(self, network=None, internet=None, gui=None):
         """Handle load of skills installed since startup."""
         if network is None:
             network = self._network_event.is_set()
         if internet is None:
             internet = self._connected_event.is_set()
+        if gui is None:
+            gui = self._gui_event.is_set() or is_gui_connected(self.bus)
 
-        self.load_plugin_skills(network=network, internet=internet)
+        # a lock is used because this can be called via state events or as part of the main loop
+        # there is a possible race condition where this handler would be executing several times otherwise
+        with self._lock:
```

Comment:
Should this same lock be referenced in `_unload_skill` to prevent a skill from being loaded and unloaded simultaneously?